File: CDTATicketing/GFSaveAddressViewController.m
//
//  GFSaveAddressViewController.m
//  CDTATicketing
//
//  Created by vishnu on 28/11/18.
//  Copyright © 2018 Genfare. All rights reserved.
//

#import "GFSaveAddressViewController.h"
#import "GFAddressTableViewCell.h"
#import "Utilities.h"
#import <MapKit/MapKit.h>

@interface GFSaveAddressViewController ()<MKLocalSearchCompleterDelegate,UITableViewDelegate,UITableViewDataSource,UITextFieldDelegate,CLLocationManagerDelegate>

@property (weak, nonatomic) IBOutlet UITextField *addressInputFld;
@property (weak, nonatomic) IBOutlet UITextField *currentLocationFld;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property (nonatomic) MKLocalSearchCompleter *searchCompleter;
@property (nonatomic) NSArray *results;
@property (nonatomic) CLLocationManager *locationManager;

@property (nonatomic) NSString *currentAddress;
@property (nonatomic) CLLocationCoordinate2D latlong;

@property (nonatomic) BOOL addressSelected;

@end

@implementation GFSaveAddressViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    self.searchCompleter = [[MKLocalSearchCompleter alloc] init];
    self.searchCompleter.delegate = self;
    self.addressInputFld.delegate = self;
    self.currentLocationFld.delegate = self;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.results = [[NSArray alloc] init];
    UINib *nib = [UINib nibWithNibName:@"GFAddressTableViewCell" bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:@"ADDRESSCELL"];
    
    [self setUpNavigationBar];
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.addressInputFld becomeFirstResponder];
    self.addressSelected = NO;
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

-(void)setUpNavigationBar {
    UIBarButtonItem *backButton = [[UIBarButtonItem alloc] initWithTitle:@"Cancel" style:UIBarButtonItemStylePlain target:self action:@selector(goBack:)];
    backButton.tintColor = UIColor.whiteColor;
    [self.navigationItem setLeftBarButtonItem:backButton];
    
    UIBarButtonItem *saveButton = [[UIBarButtonItem alloc] initWithTitle:@"Save" style:UIBarButtonItemStylePlain target:self action:@selector(saveAddress)];
    saveButton.tintColor = UIColor.whiteColor;
    [self.navigationItem setRightBarButtonItem:saveButton];
    
    if ([self.addressFor isEqualToString:KEY_SAVED_ADDRESS_HOME]) {
        self.title = @"Home Address";
    }else if([self.addressFor isEqualToString:KEY_SAVED_ADDRESS_SCHOOL]){
        self.title = @"School Address";
    }else if ([self.addressFor isEqualToString:KEY_SAVED_ADDRESS_WORK]){
        self.title = @"Work Address";
    }
}

-(void)saveAddress {
    if (!self.addressSelected) {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Alert"
                                                                       message:@"Please select an address from the list."
                                                                preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action) {}];
        
        [alert addAction:defaultAction];
        [self presentViewController:alert animated:YES completion:nil];
        
        return;
    }
    
    [Utilities saveAddress:self.currentAddress
                       lat:[NSString stringWithFormat:@"%f",self.latlong.latitude]
                      long:[NSString stringWithFormat:@"%f",self.latlong.longitude]
                       for:self.addressFor];
    [self.navigationController popViewControllerAnimated:YES];
}

-(void)showSuggessions:(NSString *)letter {
    NSString *query = [NSString stringWithFormat:@"%@%@",self.addressInputFld.text,letter];
    [self.searchCompleter setQueryFragment:query];
}

#pragma mark - UITextField Delegate methods

-(BOOL)textFieldShouldBeginEditing:(UITextField *)textField {
    if (textField == self.currentLocationFld) {
        NSLog(@"Use current Location");
        [self getUserCurrentLocation];
        [self.view endEditing:YES];
        return NO;
    }
    return YES;
}

-(BOOL)textFieldShouldReturn:(UITextField *)textField {
    return YES;
}

-(BOOL)textFieldShouldClear:(UITextField *)textField {
    [self showSuggessions:@""];
    self.addressSelected = NO;
    return YES;
}

-(BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    [self showSuggessions:string];
    return YES;
}

#pragma mark - UITableView DataSource methods

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.results.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    GFAddressTableViewCell *cell = (GFAddressTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"ADDRESSCELL"];
    MKLocalSearchCompletion *cellData = self.results[indexPath.row];
    cell.titleLabel.text = cellData.title;
    cell.descriptionLabel.text = cellData.subtitle;
    return cell;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 56;
}

#pragma mark - UITableView Delegate methods

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    MKLocalSearchRequest *searchRequest = [[MKLocalSearchRequest alloc] initWithCompletion:self.results[indexPath.row]];
    MKLocalSearchCompletion *complete = self.results[indexPath.row];
    [self showProgressDialog];
    MKLocalSearch *localSearch = [[MKLocalSearch alloc] initWithRequest:searchRequest];
    [localSearch startWithCompletionHandler:^(MKLocalSearchResponse * _Nullable response, NSError * _Nullable error) {
        if (!error) {
            for (MKMapItem *mapItem in [response mapItems]) {
                NSLog(@"%@",mapItem);
                dispatch_async(dispatch_get_main_queue(), ^{
                    self.addressInputFld.text = [NSString stringWithFormat:@"%@, %@",complete.title,complete.subtitle];
                    self.currentAddress = self.addressInputFld.text;
                    self.latlong = mapItem.placemark.coordinate;
                });
                self.addressSelected = YES;
            }
        }else{
            NSLog(@"Search request Error - %@",[error localizedDescription]);
        }
        [self dismissProgressDialog];
    }];
}

#pragma mark - Address complete delegate
-(void)completerDidUpdateResults:(MKLocalSearchCompleter *)completer {
    self.results = [completer.results copy];
    [self.tableView reloadData];
}

-(void)completer:(MKLocalSearchCompleter *)completer didFailWithError:(NSError *)error {
    NSLog(@"Handle places Error cases");
}

#pragma mark - Location based services

-(void)getUserCurrentLocation {
    _locationManager = [CLLocationManager new];
    _locationManager.delegate = self;
    _locationManager.distanceFilter = kCLDistanceFilterNone;
    _locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    
    if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 8.0 &&
        [CLLocationManager authorizationStatus] != kCLAuthorizationStatusAuthorizedWhenInUse
        //[CLLocationManager authorizationStatus] != kCLAuthorizationStatusAuthorizedAlways
        ) {
        // Will open an confirm dialog to get user's approval
        [_locationManager requestWhenInUseAuthorization];
        //[_locationManager requestAlwaysAuthorization];
    } else {
        [_locationManager startUpdatingLocation]; //Will update location immediately
    }
    [self showProgressDialog];
}

-(void)updateLocationAddressWithLocation:(CLLocation *)location {
    //TODO - update address in
    CLGeocoder *geoCoder = [[CLGeocoder alloc] init];
    [self showProgressDialog];
    [geoCoder reverseGeocodeLocation:location completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
        [self updateAddressToGFLocation:placemarks.firstObject];
        [self dismissProgressDialog];
    }];
}

-(void)updateAddressToGFLocation:(CLPlacemark *)placeMark {
    NSString *address = [NSString stringWithFormat:@"%@, %@, %@, %@, %@",
                         placeMark.name,
                         placeMark.subThoroughfare,
                         placeMark.thoroughfare,
                         placeMark.locality,
                         placeMark.ISOcountryCode];
    self.addressInputFld.text = address;
    self.currentAddress = self.addressInputFld.text;
    self.latlong = placeMark.location.coordinate;
    self.addressSelected = YES;
}

#pragma mark - Location manager delegate

- (void)locationManager:(CLLocationManager*)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    switch (status) {
        case kCLAuthorizationStatusNotDetermined: {
            NSLog(@"Location Permission is not granted yet");
        } break;
        case kCLAuthorizationStatusDenied: {
            NSLog(@"Denied Location Access");
        } break;
        case kCLAuthorizationStatusAuthorizedWhenInUse:
        case kCLAuthorizationStatusAuthorizedAlways: {
            [_locationManager startUpdatingLocation]; //Will update location immediately
        } break;
        default:
            break;
    }
}

- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray *)locations {
    CLLocation *location = [locations lastObject];
    NSLog(@"lat%f - lon%f", location.coordinate.latitude, location.coordinate.longitude);
    [self updateLocationAddressWithLocation:location];
    [_locationManager stopUpdatingLocation];
    [self dismissProgressDialog];
}

@end

File: Genfare/ViewControllers/Account/GFSaveAddressViewController.swift
//
//  GFSaveAddressViewController.swift
//  Genfare
//
//  Created by omniwyse on 12/02/19.
//  Copyright © 2019 Omniwyse. All rights reserved.
//

import UIKit
import MapKit
import CoreData

class GFSaveAddressViewController: GFBaseViewController,MKLocalSearchCompleterDelegate,UITextFieldDelegate,UITableViewDelegate,UITableViewDataSource,CLLocationManagerDelegate{
    @IBOutlet var navigationBtn: UIImageView!
    
    @IBOutlet var currentLbl: UILabel!
    @IBOutlet var bgview: UIView!
    @IBOutlet var addressTableview: UITableView!
    
     var locationManager:CLLocationManager!
    var addressFor:String = ""
    var selectedplacemark:CLPlacemark!
    var searchCompleter = MKLocalSearchCompleter()
    var searchResults = [MKLocalSearchCompletion]()
     var selectedresponse:MKLocalSearch.Response?
    @IBOutlet var addressInputField: UITextField!
    @IBOutlet var currentLocationField: UITextField!
    override func viewDidLoad() {
        super.viewDidLoad()
         self.addressInputField.delegate = self;
        barButtonIteams()
     currentLocationField.addTarget(self, action: #selector(getCurrentLocation), for: UIControlEvents.touchDown)
        self.searchCompleter.delegate = self
        // Do any additional setup after loading the view.
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        view.backgroundColor = UIColor.formLabelText
        navigationController?.setNavigationBarHidden(false, animated: false);
        navigationController?.navigationBar.barTintColor = UIColor.buttonBGBlue
            }
    func showSuggessions(_ letter: String?) {
        let query = "\(String(describing: addressInputField.text!))\(letter!)"
        searchCompleter.queryFragment = query
    }
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        showSuggessions(string)
        return true
    }
    
    func barButtonIteams(){
        var cancelBarBtn = UIBarButtonItem(
            title: "Cancel",
            style: .plain,
            target: self,
            action: #selector(cancelClicked)
        )
        self.navigationItem.leftBarButtonItem = cancelBarBtn
       
    var saveBarBtn = UIBarButtonItem(
            title: "save",
            style: .plain,
            target: self,
            action: #selector(saveClicked)
        )
        var addressTitle = addressFor.uppercased() + "ADDRESS"
        self.navigationItem.title = addressTitle
        self.navigationItem.rightBarButtonItem  = saveBarBtn
         self.navigationController?.navigationBar.tintColor = UIColor.white
    }
   
    
    @objc func getCurrentLocation(textField: UITextField) {
       determineCurrentLocation()
    }
    @objc func cancelClicked(sender: UIBarButtonItem) {
          navigationController?.popViewController(animated: true)
    }
    @objc func saveClicked(sender: UIBarButtonItem) {
        if(addressInputField.text == nil){
       popupAlert(title: "Alert", message: "Please select an address from the list", actionTitles: ["OK"], actions: [nil])
        }
        else{
        if(selectedplacemark != nil){
            savecompleteAddress(response: selectedplacemark)
        }else{
            print("no")
        }
        }
    }
    
    func completerDidUpdateResults(_ completer: MKLocalSearchCompleter) {
        searchResults = completer.results
        addressTableview.reloadData()
    }
    
    func completer(_ completer: MKLocalSearchCompleter, didFailWithError error: Error) {
        // handle error
    }
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return searchResults.count
    }
    
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let searchResult = searchResults[indexPath.row]
        let cell = addressTableview.dequeueReusableCell(withIdentifier: "GFAddressTableViewCell", for: indexPath) as! GFAddressTableViewCell
        cell.titleLabel.text = searchResult.title
        cell.descriptionLabel.text = searchResult.subtitle
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let completion = searchResults[indexPath.row]
        
        let searchRequest = MKLocalSearchRequest(completion: completion)
        self.addressInputField.text = completion.title   + completion.subtitle
        let search = MKLocalSearch(request: searchRequest)
        search.start { (response, error) in
             let placemark = response?.mapItems[0].placemark
            self.selectedplacemark = placemark!
            self.selectedresponse = response
        }
    }
    
    func savecompleteAddress(response:CLPlacemark){
        do{
        let placemark = response
       
        let name = placemark.name
        let sub = placemark.subThoroughfare
        let thoroughfare = placemark.thoroughfare
        let iso = placemark.isoCountryCode
        
        let managedContext = GFDataService.context
        let storedContents = NSEntityDescription.entity(forEntityName: "StoredAddress", in: managedContext)
        var storeObj:StoredAddress
        let fetchRequest:NSFetchRequest = StoredAddress.fetchRequest()
            if(addressFor as? String == "home" ){
        fetchRequest.predicate = NSPredicate(format: "type == %@",("home"))
            }
            else if(addressFor as? String == "work"){
                fetchRequest.predicate = NSPredicate(format: "type == %@",("work"))
            }
            else if(addressFor as? String == "school"){
                fetchRequest.predicate = NSPredicate(format: "type == %@",("school"))
            }
        let fetchResults = try managedContext.fetch(fetchRequest) as! Array<StoredAddress>
        if fetchResults.count <= 0 {
             storeObj = NSManagedObject(entity: storedContents!, insertInto: managedContext) as! StoredAddress
        }else{
            storeObj = fetchResults.first!
        }
        
        storeObj.name = name
        storeObj.subThoroughfare = sub
        storeObj.thoroughfare = thoroughfare
        storeObj.isoCountryCode = iso
        storeObj.type = addressFor as! String
        }
        catch{
            print("saving failed ")
        }

        GFDataService.saveContext()
        navigationController?.popViewController(animated: true)
    }
    func determineCurrentLocation()
    {
        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.requestAlwaysAuthorization()
        
        if CLLocationManager.locationServicesEnabled() {
           
            locationManager.startUpdatingLocation()
        }
    }
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        let userLocation:CLLocation = locations[0] as CLLocation
        
        let geocoder = CLGeocoder()
        geocoder.reverseGeocodeLocation(userLocation) { (placemarks, error) in
            if (error != nil){
                print("error in reverseGeocode")
            }
            let placemark = placemarks! as [CLPlacemark]
            if placemark.count>0{
                let placemark = placemarks![0]
                self.selectedplacemark = placemark
                self.currentLocationField.isUserInteractionEnabled = false
                let name = placemark.name
               let thoroughfare = placemark.thoroughfare
                self.addressInputField.text = name! + thoroughfare!
                
               
            }
        }
        
       
    }
    
}

